lib/listen/record.rb

```diff
@@ -32,27 +27,20 @@ def unset_path(dir, rel_path)
     def file_data(dir, rel_path)
       root = @paths[dir.to_s]
       dirname, basename = Pathname(rel_path).split.map(&:to_s)
-      if [nil, '', '.'].include? dirname
-        root[basename] ||= {}
-        root[basename].dup
-      else
-        root[dirname] ||= {}
-        root[dirname][basename] ||= {}
-        root[dirname][basename].dup
-      end
+      dirname = '.' if [nil, '', '.'].include? dirname
+      root[dirname] ||= {}
+      root[dirname][basename] ||= {}
+      root[dirname][basename].dup
     end
 
     def dir_entries(dir, rel_path)
-      tree = if [nil, '', '.'].include? rel_path.to_s
-               @paths[dir.to_s]
-             else
-               @paths[dir.to_s][rel_path.to_s] ||= _auto_hash
-               @paths[dir.to_s][rel_path.to_s]
-             end
+      rel_path = '.' if [nil, '', '.'].include? rel_path.to_s
+      @paths[dir.to_s][rel_path.to_s] ||= _auto_hash
+      tree = @paths[dir.to_s][rel_path.to_s]
 
       result = {}
       tree.each do |key, values|
-        # only get data for file entries
+        # only get data for file/dir entries
         result[key] = values.key?(:mtime) ? values : {}
```

Comment:
To elaborate I believe it should be:

'.' contains [x,y,z,.git]
'.git' contains [a,b,c]
So folder names appear as key pointing to contents, but also within contents of their parent. Of course '.' is root with no parent.